### DISCUSSION TITLE
Examples depend on "using namespace ace_button" in the global namespace

### DISCUSSION THREAD

**Discussion by drf5n** [bxparks/AceButton#79] [General]
I found that https://github.com/bxparks/AceButton#IncludeHeader explained how to import the ace_button namespace into global, but the examples seem to all demonstrate dependency on the ace_button:: members all being in the global namespace as being best practice.

Shouldn't the recommended uses be more encapsulated? Like:

```
/*
 * A demo of a simplest AceButton that has a visible effect. One button is
 * connected to the digital pin BUTTON_PIN. It uses the internal pull-up
 * resistor (INPUT_PULLUP). Pressing the button turns on the built-in LED.
 * Releasing the button turns off the LED.
 */

#include <AceButton.h>

// Some ESP32 boards have multiple builtin LEDs so don't define LED_BUILTIN.
#if defined(ESP32)
  const int LED_PIN = 2;
#else
  const int LED_PIN = LED_BUILTIN;
#endif

const int BUTTON_PIN = 2;
const int LED_ON = HIGH;
const int LED_OFF = LOW;

ace_button::AceButton button(BUTTON_PIN);

// Forward reference to prevent Arduino compiler becoming confused.
void handleEvent(ace_button::AceButton*, uint8_t, uint8_t);

void setup() {
  delay(2000);

#if defined(ARDUINO_AVR_LEONARDO)
  RXLED0; // LED off
  TXLED0; // LED off
#endif

  pinMode(LED_PIN, OUTPUT);
  pinMode(BUTTON_PIN, INPUT_PULLUP);
  button.setEventHandler(handleEvent);
}

void loop() {
  button.check();
}

void handleEvent(ace_button::AceButton* /* button */, uint8_t eventType,
    uint8_t /* buttonState */) {
  using namespace ace_button;
  switch (eventType) {
    case AceButton::kEventPressed:
      digitalWrite(LED_PIN, LED_ON);
      break;
    case AceButton::kEventReleased:
      digitalWrite(LED_PIN, LED_OFF);
      break;
  }
}
```

... or with a specific name import instead of the wildcard:

```
/*
 * A demo of a simplest AceButton that has a visible effect. One button is
 * connected to the digital pin BUTTON_PIN. It uses the internal pull-up
 * resistor (INPUT_PULLUP). Pressing the button turns on the built-in LED.
 * Releasing the button turns off the LED.
 */

#include <AceButton.h>
using ace_button::AceButton;

// Some ESP32 boards have multiple builtin LEDs so don't define LED_BUILTIN.
#if defined(ESP32)
  const int LED_PIN = 2;
#else
  const int LED_PIN = LED_BUILTIN;
#endif

const int BUTTON_PIN = 2;
const int LED_ON = HIGH;
const int LED_OFF = LOW;

AceButton button(BUTTON_PIN);

// Forward reference to prevent Arduino compiler becoming confused.
void handleEvent(AceButton*, uint8_t, uint8_t);

void setup() {
  delay(2000);

#if defined(ARDUINO_AVR_LEONARDO)
  RXLED0; // LED off
  TXLED0; // LED off
#endif

  pinMode(LED_PIN, OUTPUT);
  pinMode(BUTTON_PIN, INPUT_PULLUP);
  button.setEventHandler(handleEvent);
}

void loop() {
  button.check();
}

void handleEvent(AceButton* /* button */, uint8_t eventType,
    uint8_t /* buttonState */) {
     // (commented out redundancy per response below) using namespace ace_button;

  switch (eventType) {
    case AceButton::kEventPressed:
      digitalWrite(LED_PIN, LED_ON);
      break;
    case AceButton::kEventReleased:
      digitalWrite(LED_PIN, LED_OFF);
      break;
  }
}
```

**Comment by bxparks** [comment]
Yes I agree that wildcard imports of namespaces (i.e. `using namespace xxx`) should be used sparingly. I use them in my AceButton examples deliberately, for several reasons:

1) C++ namespaces is a somewhat advanced topic, and I figure that many, if not most, Arduino programmers are struggling with a lot of basic C and C++ concepts, and the multitude of ways that C++ allows someone to shoot themselves in the foot. I want to keep the examples as simple as possible, without extraneous clutter.
2) I figure that the people who are aware of the problems of namespace pollution are the more advanced users who will know how to navigate the namespace issues.
3) I make a distinction between an end-user application program versus a library. Within my library, more precisely, within a publicly exported header file in my libraries, I never use the `using` or `using namespace` statement, exactly to prevent polluting the namespace. However, in an end-user application (i.e. the `*.ino` file), which is at the leaf of the dependency tree, I think it is acceptable to use a `using` or `using namespace` statement, because the end-user has the ability to remove those statement and fix any name collisions.
4) Almost all 3rd party Arduino libraries are written in the global namespace. Only a very tiny fraction use namespaces, like my libraries. I figure that my examples with the `using namespace ace_button` are no worse than the vast majority of other Arduino libraries. But at least with my libraries, the end-user has the **capability** of fixing the name collisions by removing the `using` statements.

With regards to your example code, your second example uses the `using ace_button::AceButton` statement, which I mention in my README.md. However, once you do that, I think you can remove the `using namespace ace_button` from inside `handleEvent()`, because the `AceButton` symbol has already been imported into the global namespace.

**Reply by drf5n** [reply]
You are correct -- in my 2nd example with the `using ace_button::AceButton`, the `using namespace ace_button` was redundant.

My namespace awareness is more from the R, Perl and Python side, and my current C/C++ is more script-kiddy.  I really like that you use the namespace feature, and I'm glad that I learned from your code.  You are correct that the vast majority of Arduino libraries do not take the care to use namespace, but that difference makes your library a little odd. That difference also makes your library easier to combine or retrofit into to other projects.

Maybe noting the namespace-respecting coding pattern in the doc at https://github.com/bxparks/AceButton#include-header-and-use-namespace would have helped me understand the advantage of the difference quicker.  For example:

----
## Include Header and Use Namespace
Only a single header file AceButton.h is required to use this library. To prevent name clashes with other libraries that the calling code may use, all classes are defined in the ace_button namespace. For example:

```
#include <AceButton.h>
ace_button::AceButton button(BUTTON_PIN);
...
```

To use the code without prepending the ace_button:: prefix, use the `using` directive to import the namespace into global:
       
```
#include <AceButton.h>
using namespace ace_button;

```
----
Thank you for your work.
